[docs] Clarify runtime update header overrides (#48535)
expo/expo: docs/pages/eas-update/override.mdx

> **warning** The feature described in this section is available in Expo SDK 54 with the `expo-updates` version 0.29.0 and later.

> For apps using EAS Update, the primary use case for this feature is [channel surfing](/eas-update/channel-surfing/), which switches the `expo-channel-name` request header at runtime as part of a complete channel switching workflow. You can also override other custom request headers if your update server supports them.

Reviewer: ✏️ **style** `mechanics.md#vale-sentencelength`

The first sentence is now 31 words: "For apps using EAS Update, the primary use case ... complete channel switching workflow." Split it, for example: "For apps using EAS Update, the primary use case for this feature is channel surfing. Channel surfing switches the `expo-channel-name` request header at runtime as part of a complete channel switching workflow."

```suggestion
For apps using EAS Update, the primary use case for this feature is [channel surfing](/eas-update/channel-surfing/). Channel surfing switches the `expo-channel-name` request header at runtime as part of a complete channel switching workflow. You can also override other custom request headers if your update server supports them.
```

<!-- docs-review -->

For example, if you have a `default` channel for production updates and a `preview` channel for preview updates, you can override the `expo-channel-name` request header to point to the `preview` channel. This enables you to test preview updates in your current production builds.

Another potential use case is to provide different updates to different users, for example, so that a group of internal users (such as employees) receive updates before end-users.

### How it works

> You can override request headers by calling [`Updates.setUpdateRequestHeadersOverride`](/versions/latest/sdk/updates/#updatessetupdaterequestheadersoverriderequestheaders). With EAS Update, overriding the `expo-channel-name` request header requests updates from the specified channel.

Reviewer: 💡 **nit** `prose.md#core-principles`

"overriding the `expo-channel-name` request header requests updates" stumbles on "header requests"; consider "makes the app request updates from the specified channel".

```suggestion
You can override request headers by calling [`Updates.setUpdateRequestHeadersOverride`](/versions/latest/sdk/updates/#updatessetupdaterequestheadersoverriderequestheaders). With EAS Update, overriding the `expo-channel-name` request header makes the app request updates from the specified channel.
```

<!-- docs-review -->

> **warning** The feature described in this section is available in Expo SDK 54 with the `expo-updates` version 0.29.0 and later.

> For apps using EAS Update, the primary use case for this feature is [channel surfing](/eas-update/channel-surfing/). Channel surfing switches the `expo-channel-name` request header at runtime as part of a complete channel switching workflow. You can also override other custom request headers if your update server supports them.

Reviewer 2: > You can also override other custom request headers if your update server supports them.

we don't support all headers. the prerequisite is that the header is defined in `updates.requestHeaders`. that prevents people from accidentally override headers like `expo-platform`

suggest to remove the sentence?

await Updates.reloadAsync();
```

> > **info** You can only override a request header if its name is included in the native build. When an EAS Build profile defines `channel`, EAS Build includes `expo-channel-name` automatically. If you build without EAS Build, add the header to [`updates.requestHeaders`](/eas-update/getting-started/#configure-update-channels-in-appjson) or the native project, then rebuild the app.

Reviewer: ✏️ **style** `prose.md#voice-and-tone`

"if its name is included in the native build" is passive. Use the active form: "if the native build includes its name".

```suggestion
> **info** You can only override a request header if the native build includes its name. When an EAS Build profile defines `channel`, EAS Build includes `expo-channel-name` automatically. If you build without EAS Build, add the header to [`updates.requestHeaders`](/eas-update/getting-started/#configure-update-channels-in-appjson) or the native project, then rebuild the app.
```

<!-- docs-review -->

> > **warning** Switching channels can cause compatibility issues when updates use different migrations or data shapes. See [Risks and considerations when switching channels](/eas-update/channel-surfing/#risks-and-considerations-when-switching-channels).

Reviewer: ✏️ **style** `mdx-components.md#callouts`

This `warning` callout now sits directly under the `info` callout, and stacked callouts of different types should not follow each other. Consider moving this switching-risks warning up beside the channel surfing mention (after the "For example, if you have a `default` channel" paragraph), or folding it into prose.

<!-- docs-review -->

await Updates.reloadAsync();
```

> > **info** You can only override a request header if the native build includes its name. When an EAS Build profile defines `channel`, EAS Build includes `expo-channel-name` automatically. If you build without EAS Build, add the header to [`updates.requestHeaders`](/eas-update/getting-started/#configure-update-channels-in-appjson) or the native project, then rebuild the app.

Reviewer 2: > EAS Build includes `expo-channel-name` automatically.

back to the customer's question. not sure if we want to clearly say what's the "channel" value we set from eas build? e.g. the channel in eas.json. if you select "production" build, the channel is "production" by default

Author: Thanks @Kudo, good point 🙏  
I updated it, do you think this is better?

> > **info** You can only override a request header if the native build includes its name. When a build profile in **eas.json** defines `channel`, EAS Build includes the `expo-channel-name` header and sets it to that `channel` value. For example, a build created with the `production` profile and `"channel": "production"` requests updates from the `production` channel. If you build without EAS Build, add the header to [`updates.requestHeaders`](/eas-update/getting-started/#configure-update-channels-in-appjson) or the native project, then rebuild the app.

Reviewer 2: ```suggestion
> **info** You can only override a request header if the native build includes its name. When an EAS Build profile defines `channel`, EAS Build includes `expo-channel-name` automatically. If you build without EAS Build, add the header to [`updates.requestHeaders`](/eas-update/getting-started/#configure-update-channels-in-appjson) or the native project, then rebuild the app.

<Collapsible summary="What's the default channel specific by EAS Build?">

When a build profile in **eas.json** defines `channel`, EAS Build includes the `expo-channel-name` header and sets it to that `channel` value. For example, a build created with the `production` profile and `"channel": "production"` in its **eas.json**. The channel would be `production` by default.
</Collapsible>
```

Maybe just split into separate Collapsible? because these are two difference things
- You can only override a request header if the native build includes its name.
- What's the default channel

please proofread and refine the sentences for my change. it just a proof-of-concept

Reviewer 3: +1 to Kudo's suggestion, our docs writing guide says the `info` blocks should have 1 short paragraph so breaking the details out to a collapsible or separate paragraph makes sense.

There are two relevant APIs:

> 1. `Updates.setUpdateURLAndRequestHeadersOverride({ updateUrl: string, requestHeaders: Object })` - this method overrides the update URL and the request headers that are specified in **app.json** / **Expo.plist** / **AndroidManifest.xml**, such as the `expo-channel-name` header.

Reviewer: ✏️ **style** `mechanics.md#no-space-around-slash`

"**app.json** / **Expo.plist** / **AndroidManifest.xml**" has spaces around the slashes. Remove them, matching the **app.json**/**app.config.js** pattern used on other EAS Update pages.

```suggestion
1. `Updates.setUpdateURLAndRequestHeadersOverride({ updateUrl: string, requestHeaders: Object })` - this method overrides the update URL and the request headers that are specified in **app.json**/**Expo.plist**/**AndroidManifest.xml**, such as the `expo-channel-name` header.
```

<!-- docs-review -->
